### DISCUSSION TITLE
StructLayoutAttribute incomplete (or broken or doco incomplete)

### DISCUSSION THREAD

**Discussion by UberKluger** [dotnet/csharplang#5580] [Language Ideas]
In attempting to answer a stack overflow question [Convert Byte Array (from legacy program data file) to Powershell object](https://stackoverflow.com/questions/40981836/convert-byte-array-from-legacy-program-data-file-to-powershell-object) I ran into several problems regarding the use of `StructLayoutAttribute`. The answer to the question seemed obvious, just add a type into the current Powershell AppDomain which has a C# static method to take the array of bytes and dump the raw bytes into a appropriate struct using a pointer to the struct (as in C or C++) then return the struct. For C# this requires the use of `unsafe` code but that should be no problem as long as you can get the struct members in the right place, i.e. correct offsets. That's when the fun began. (Note: I have since realised that this is not a general solution to the SO question since it assumes that the incoming data will have the same alignment requirements as the final struct. Not guaranteed.)

My initial floundering as I iterated through research a bit -> write a bit -> compile & fail -> research a bit ... has been skipped over. Finally got to something that looked right and compiled with the correct layout
```
[StructLayout(LayoutKind.Sequential)]
public unsafe struct test {
 	public int	anint; 	// define this as offset 0
	public short	short1; 	// offset 4 OK
	public short	short2; 	// offset 6 looks promising
	public fixed byte	byte_arr[4];	// equivalent to C array of byte (unsigned char). offset 8 nearly there
	public fixed char	char_arr[10]; 	// offset 12 just right
};
```
Ran some tests (checking offsets of members as noted above), seems fine then tried to load in some data BUT when checking the results found that the "arrays" weren't (arrays). Instead, Powershell says that they are a (compiler) manufactured type and only the first element is directly accessible (as the member "FixedElementField"). Turns out, within a C# unsafe context, the struct member (e.g. `byte_arr`) is evaluated as a pointer to the first element which is indexed to access the other elements. No good for Powershell. Have to use normal array objects somehow. Try to include them in the same object to avoid cluttering the Powershell variable space (or returning a multi-level object).
```
[StructLayout(LayoutKind.Sequential)]
unsafe struct test {
 	public int	anint; 
	public short	short1;
	public short	short2;
	public fixed byte	byte_arr[4];
	public fixed char	char_arr[10];
	public byte[]	thebytes;
	public char[]	thechars;
};
```
Just have to copy from `byte_arr` to `thebytes`, only slightly inconvenient but, WHAT THE HEY, now testing shows that `char_arr` is no longer at offset 12. It's been realigned on an 8 byte boundary (offset 16). Even setting `[StructLayout(LayoutKind.Sequential, Pack=1)]` won't fix it. Well that's just annoying. Now I can't just pump my `byte` array into the struct. No mention of this behaviour in the `StructLayoutAttribute`, `LayoutKind.Sequential` or `fixed` documentation. Why would including managed members mess things up? Because they're 8 bytes in size? Try this:
```
[StructLayout(LayoutKind.Sequential)]
unsafe struct test {
 	public int	anint; 
	public short	short1;
	public short	short2;
	public fixed byte	byte_arr[4];
	public fixed char	char_arr[10];
	public double	thebytes;
};
```
Nope, not the reason. `char_arr` is back at offset 12. `thebytes` is at 32 (12 + 2 * 10). What a pain. Oh well, maybe I can force the alignment using `LayoutKind.Explicit`.
```
[StructLayout(LayoutKind.Explicit)]
unsafe struct test {
 	[FieldOffset(0)] public int	anint; 
	[FieldOffset(4)] public short	short1;
	[FieldOffset(6)] public short	short2;
	[FieldOffset(8)] public fixed byte	byte_arr[4];
	[FieldOffset(12)] public fixed char	char_arr[10];
	[FieldOffset(32)] public byte[]	thebytes;
	[FieldOffset(?)] public char[]	thechars;	// Um, how big is a reference.
};
```
Earlier, I assumed that a reference takes 8 bytes but that's just an assumption. Several sources seem to concur but they are still either opinions or based on a particular implementation. As far as I can determine from the ***official literature***, no size should be used generally for a reference. There isn't even a safe way to measure it. Can't add a reference to a struct and measure the size change because it becomes a "managed" object and `sizeof()` is not allowed. Tried putting an addressable member on either side but that really went south.
```
[StructLayout(LayoutKind.Sequential)]
unsafe struct test {
 	public int	anint; 
	public short	short1;
	public short	short2;
	public fixed byte	byte_arr[4];
	public fixed char	char_arr[10];
	public char[]	thechars;
	public double	thebytes;
};
```
`thebytes` ended up at offset -8(!?!) relative to `anint` so not only could I not see the size of `thechars` but the sequential ordering got messed up (even more). Just for giggles (which I really needed at this point), I tried the following to see how badly it was screwed up.
```
[StructLayout(LayoutKind.Sequential)]
unsafe struct test {
	public int	anint;
	public short	short1;
	public short	short2;
	public byte	abyte1;
	public byte	abyte2;
	public char	achar;
	public char[]	thechars;
	public double	thebytes;
};
```
And the results are in...
```
short1 @ 4, short2 @ 6
abyte1 @ 10, abyte2 @ 11
achar @ 8
thebytes @ -8
(all relative to anint, optimal packing but not sequential as directed)
```
Yep, they're coming to take me away, ha ha.

That's the story so far. To summarise (TLDR),

When trying to do low level structure member offset mapping:
1) Including managed members corrupts close packing of fixed buffers for `LayoutKind.Sequential`.
2) It also destroys the sequential ordering even without fixed buffers.
3) There is no generally safe way to have more than one managed member for `LayoutKind.Explicit` since the size of a reference is not defined (or is it?).

Disclaimer: all of the above was determined on a base Win10 system using included csc.exe for .NET 4.0.30319 (C#5.0). It is possible that these issues are long since resolved, though C#5.0 appears to be the current documentation (from MS and ECMA). Didn't see anything relevant in the headings of the list of "features" for C#6.0-10.0. Documentation for `StructLayout` was from .NET6. Tried checking open issues but I couldn't think of a useful filtering method to find the relevant ones out of 423 (filtering on "StructLayout" left 13 with none similar and "fixed" gave 96 and was too general). I could just have totally misunderstood what I was attempting and wiser heads might shake at my abject stupidity. At the very least, the confusion I have shown here could be avoided with some documentation updates (as in "Don't try this, it doesn't work").

**Comment by svick** [comment]
This is the documented behavior, but maybe a bit non-obvious. [The documentation says:](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.structlayoutattribute?view=net-6.0#remarks)

> For blittable types, `LayoutKind.Sequential` controls both the layout in managed memory and the layout in unmanaged memory. For non-blittable types, it controls the layout when the class or structure is marshaled to unmanaged code, but does not control the layout in managed memory.

Since you're interested in that layout in managed memory, `LayoutKind.Sequential` only works if the type is [blittable](https://docs.microsoft.com/en-us/dotnet/framework/interop/blittable-and-non-blittable-types). But by adding array fields, you're making the type non-blittable, which means `LayoutKind.Sequential` no longer applies.

One approach that should work is exposing the arrays as computed properties, with no backing fields. This way the type stays blittable (no array fields), but PS will see what it understands (array properties).

This would require allocating and copying an array on every access. If that's a problem, you could use a `Span<T>` property instead, but I don't know if PS supports that.

**Comment by acaly** [comment]
Related: dotnet/runtime#59593. From the discussion, it seems that the layout algorithm has some very old code inherited from .NET Framework. Hopefully it will be fixed soon.

Also, besides how the runtime handles this, I think the compiler should also try to generate errors when this attribute is applied to non-blittable structs.

**Reply by HaloFour** [reply]
> Also, besides how the runtime handles this, I think the compiler should also try to generate errors when this attribute is applied to non-blittable structs.

Sounds like a use case for an analyzer, especially since that attribute still does apply to marshalling structs that aren't blittable.
